src/testRunner/unittests/tsserver/helpers.ts

```diff
@@ -55,46 +55,80 @@ namespace ts.projectSystem {
         readonly callback: TI.RequestCompletedAction;
     }
 
-    export const nullLogger: server.Logger = {
-        close: noop,
-        hasLevel: returnFalse,
-        loggingEnabled: returnFalse,
-        perftrc: noop,
-        info: noop,
-        msg: noop,
-        startGroup: noop,
-        endGroup: noop,
-        getLogFileName: returnUndefined,
-    };
+    export interface Logger extends server.Logger {
+        logs: string[];
+    }
 
-    export function createHasErrorMessageLogger() {
-        let hasErrorMsg = false;
-        const { close, hasLevel, loggingEnabled, startGroup, endGroup, info, getLogFileName, perftrc } = nullLogger;
-        const logger: server.Logger = {
-            close, hasLevel, loggingEnabled, startGroup, endGroup, info, getLogFileName, perftrc,
-            msg: (s, type) => {
-                Debug.fail(`Error: ${s}, type: ${type}`);
-                hasErrorMsg = true;
-            }
+    export function nullLogger(): Logger {
+        return {
+            close: noop,
+            hasLevel: returnFalse,
+            loggingEnabled: returnFalse,
+            perftrc: noop,
+            info: noop,
+            msg: noop,
+            startGroup: noop,
+            endGroup: noop,
+            getLogFileName: returnUndefined,
+            logs: [],
         };
-        return { logger, hasErrorMsg: () => hasErrorMsg };
     }
 
-    export function createLoggerWritingToConsole(): server.Logger {
-        const { close, startGroup, endGroup, getLogFileName } = nullLogger;
+    export function createHasErrorMessageLogger(): Logger {
         return {
-            close,
+            ...nullLogger(),
+            msg: (s, type) => Debug.fail(`Error: ${s}, type: ${type}`),
+        };
+    }
+
+    export function createLoggerWritingToConsole(): Logger {
+        return {
+            ...nullLogger(),
             hasLevel: returnTrue,
             loggingEnabled: returnTrue,
             perftrc: s => console.log(s),
             info: s => console.log(s),
             msg: (s, type) => console.log(`${type}:: ${s}`),
-            startGroup,
-            endGroup,
-            getLogFileName
         };
     }
 
+    export function createLoggerWithInMemoryLogs(): Logger {
+        const logger = createHasErrorMessageLogger();
+        return {
+            ...logger,
+            hasLevel: returnTrue,
+            loggingEnabled: returnTrue,
+            info: s => logger.logs.push(
+                s.replace(/Elapsed::?\s*\d+(?:\.\d+)?ms/g, "Elapsed:: *ms")

```

Comment:
Is there some way we can detect that a baseline change includes something that looks like a duration and direct the author to this code?  It seems like adding new log statements could be confusing otherwise (accept baselines locally, fail in CI, re-invent sanitization?).

Reply:
cant think of anything.. we do this in multiple places but in most cases i have realized it as part of just running tests